### PR TITLE
android(feat):new arch support

### PR DESCRIPTION
**Pull request by gosha212**
## Description

- This pull request addresses the issue described here: #4175

In this pull request, I have added support for the new arch in detox

---

> _For features/enhancements:_
 - [ ] I have added/updated the relevant references in the [documentation](https://github.com/wix/Detox/tree/master/docs) files.

> _For API changes:_
 - [ ] I have made the necessary changes in the [types index](https://github.com/wix/Detox/blob/master/detox/index.d.ts) file.

### CODE REVIEW

File: .buildkite/jobs/pipeline.android_demo_app_rn_76.yml [hunk 1]
```diff
@@ -1,4 +1,4 @@
-  - label: ":android::react: RN .76 + Android: Demo app"
+  - label: ":android::react: (New Arch) - RN .76 + Android: Demo app"
```
Comment (asafkorem) [inline .buildkite/jobs/pipeline.android_demo_app_rn_76.yml:1]: Why not testing both new arch and old arch on .76?
Is that a breaking change?

**Comment by asafkorem** [comment]
@gosha212 don't you crash on the `LanguageScreen`?

**Comment by asafkorem** [comment]
we should check if our bridge native modules work with backward comp. or if there's something magically /hybrid that happens here

File: .buildkite/jobs/pipeline.android_demo_app_rn_76.yml (hunk 1 above)
Comment (d4vidi) [reply .buildkite/jobs/pipeline.android_demo_app_rn_76.yml:1]: I think it's ok for the demo app but the question is valid for the main test app; By now we know the backwards compatibility is feasible with a reasonable effort so let's take it

File: detox/test/e2e/33.attributes.test.js [hunk 2]
```diff
@@ -143,7 +143,8 @@ describe('Attributes', () => {
       });
     });
 
-    it(':android: should have a boolean .value', async () => {
+    // Checkbox is not working with the new arch yet
+    it.skip(':android: should have a boolean .value', async () => {
```
Comment (asafkorem) [inline detox/test/e2e/33.attributes.test.js:147]: we should not skip this test for the legacy if it works there
let's add the skip marker for `@new-arch` and `@legacy` on `rnSkipper`.

**Review by asafkorem — CHANGES_REQUESTED** [review]
In iOS we can check the env flag `process.env.RCT_NEW_ARCH_ENABLED`
We can use this when running new arch on android as well (for the tests)

File: detox/test/e2e/33.attributes.test.js (hunk 2 above)
Comment (gosha212) [reply detox/test/e2e/33.attributes.test.js:147]: I think it will be easier to contribute to checkbox project =)

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/ReactNativeLoadingMonitor.kt [hunk 3]
```diff
@@ -79,51 +76,34 @@ open class ReactNativeLoadingMonitor(
...
-            else -> Any()
+    private fun isReactNativeLoaded(reactContext: ReactContext?) =
+        reactContext != null && reactContext !== previousReactContext && reactContext.hasActiveReactInstance()
+
+    private fun subscribeAsyncRNContextHandler(onReactContextInitialized: () -> Any) {
+        val isFabric = isFabricEnabled()
+        if (isFabric) {
+            // We do a casting for supporting RN 0.75 and above
+            val host = rnApplication.reactHost as ReactHostImpl?
+            host?.addReactInstanceEventListener(object : ReactInstanceEventListener {
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/ReactNativeLoadingMonitor.kt:90]: `host?` is an issue... If we reach this subscribe() flow and host is null, we'd stay stuck forever

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/ReactNativeLoadingMonitor.kt [hunk 4]
```diff
@@ -69,8 +64,10 @@ open class ReactNativeLoadingMonitor(
                 // Due to an ugly timing issue in RN
                 // it is possible that our listener won't be ever called
                 // That's why we have to check the reactContext regularly.
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/ReactNativeLoadingMonitor.kt:66]: Can we Detox this comment a bit? 😄 Is it really an _ugly_ unexplained timing issue or do we understand things better now and give a good technical explanation (for future generations' sake)

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/reloader/ReactNativeReloader.kt [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+class NewArchitectureNativeReLoader(
+    private val instrumentation: Instrumentation,
+    private val rnApplication: ReactApplication
+) : ReactNativeReLoader {
+    override fun reloadInBackground() {
+        val reactHost = rnApplication.reactHost
+        instrumentation.runOnMainSync {
+            reactHost?.reload("Detox")
+        }
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/reloader/ReactNativeReloader.kt:31]: Here too, if `reactHost` is null then reload would just fail silently

Comment (d4vidi) [reply detox/android/detox/src/full/java/com/wix/detox/reactnative/reloader/ReactNativeReloader.kt:31]: Regardless, maybe this would be a (little) bit cleaner:
```suggestion
        instrumentation.runOnMainSync {
            rnApplication.reactHost?.reload("Detox")
        }
```
I mean, `rnApplication` is a class member, so no need to hold an additional local var no top of that

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/timers/FabricTimersIdlingResource.kt [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        getChoreographer().postFrameCallback(this)
+    }
+
+    override fun onUnregistered() {
+        super.onUnregistered()
+        getChoreographer().removeFrameCallback(this)
+    }
+
+    override fun getDebugName(): String {
+        return "FabricTimers"
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/timers/FabricTimersIdlingResource.kt:49]:
```suggestion
        return "timers"
```
We don't want "Fabric" in the sync-debug logs

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/timers/FabricTimersIdlingResource.kt [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+class FabricTimersIdlingResource(
+    private val reactContext: ReactContext,
+    private val getChoreographer: () -> Choreographer = { Choreographer.getInstance() }
+) : DetoxIdlingResource(), Choreographer.FrameCallback {
+
+    override fun checkIdle(): Boolean {
+        val hasActiveTimers: Boolean = hasActiveTimers()
+
+        if (hasActiveTimers) {
+            getChoreographer().postFrameCallback(this@FabricTimersIdlingResource)
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/timers/FabricTimersIdlingResource.kt:21]: Future note (todo?): Add ourselves as a callback in the timers module, in-sync with the longest current wait in queue*. Brut-force checking every 16ms was not the best idea, in retrospect.

_\* still below the window threshold..._

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/timers/FabricTimersIdlingResource.kt [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+        return !hasActiveTimers
+    }
+
+    private fun hasActiveTimers(): Boolean {
+        val timersManager = getTimersManager()
+        val hasActiveTimersInRangeInstanceClass = timersManager::class
+        val method = hasActiveTimersInRangeInstanceClass.declaredFunctions.first { it.name.contains("hasActiveTimersInRange") }
+        method.isAccessible = true
+        val hasActiveTimers: Boolean = method.call(timersManager, BUSY_WINDOW_THRESHOLD) as? Boolean ?: false
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/timers/FabricTimersIdlingResource.kt:34]: Why not use Reflect for this as well?

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/timers/FabricTimersIdlingResource.kt [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    override fun getBusyHint(): Map<String, Any>? = null
+
+    override fun getName(): String {
+        return FabricTimersIdlingResource::class.java.name
+    }
+
+    private fun getTimersManager(): JavaTimerManager {
+        val reactHost = Reflect.on(reactContext).field("mReactHost").get<Any>()
+        val reactInstance = Reflect.on(reactHost).field("mReactInstance").get<Any>()
+        return Reflect.on(reactInstance).field("mJavaTimerManager").get() as JavaTimerManager
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/timers/FabricTimersIdlingResource.kt:61]: Could we outsource all reflection ugliness elsewhere? I want to say a `TimersManagerReflected` kind of wrapper but my idea is as good as yours (as long as this is put away from the core logic which this class holds)

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/reloader/ReactNativeReloader.kt (hunk 5 above)
Comment (d4vidi) [reply detox/android/detox/src/full/java/com/wix/detox/reactnative/reloader/ReactNativeReloader.kt:31]: Anyways, nice that it's a simple `reactHost.reload()` call now 😄

**Review by d4vidi — CHANGES_REQUESTED** [review]
Overall, great job. Lot's of notes on my end but we are nevertheless very close to finish 👑

File: detox/android/detox/build.gradle [hunk 10]
```diff
@@ -147,6 +147,12 @@ dependencies {
     api('androidx.test.uiautomator:uiautomator:2.2.0') {
         because 'Needed by Detox but also makes UIAutomator seamlessly provided to Detox users with hybrid apps/E2E-tests.'
     }
+    api('androidx.test:core-ktx:1.6.1') {
+        because 'Needed by Detox but also makes AndroidX test core seamlessly provided to Detox users with hybrid apps/E2E-tests.'
+    }
+    api("org.jetbrains.kotlin:kotlin-reflect:$_kotlinVersion") {
+        because('Needed by Detox for kotlin reflection')
+    }
```
Comment (d4vidi) [inline detox/android/detox/build.gradle:155]: on 2nd thought, maybe `implementation` would be a better idea here

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/ReactNativeLoadingMonitor.kt (hunk 3 above)
Comment (d4vidi) [reply detox/android/detox/src/full/java/com/wix/detox/reactnative/ReactNativeLoadingMonitor.kt:90]: I see but here the code implicitly suggests that adding the listener is just "nice-to-have", maybe `host!!` would make it more explicit

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/factory/FabricDetoxIdlingResourceFactoryStrategy.kt [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+
+class FabricDetoxIdlingResourceFactoryStrategy(private val reactApplication: ReactApplication) :
+    DetoxIdlingResourceFactoryStrategy {
+    override suspend fun create(): Map<IdlingResourcesName, DetoxIdlingResource> =
+        withContext(Dispatchers.Main) {
+            val reactContext =
+                reactApplication.getCurrentReactContext()
+                    ?: throw IllegalStateException("ReactContext is null")
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/factory/FabricDetoxIdlingResourceFactoryStrategy.kt:20]: So maybe:
```suggestion
                reactApplication.getCurrentReactContextSafe()
```

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/uimodule/fabric/FabricUIManagerIdlingResources.kt [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    override fun onUnregistered() {
+        super.onUnregistered()
+        Choreographer.getInstance().removeFrameCallback(this)
+    }
+
+    override fun getDebugName(): String {
+        return "Fabric UI"
+    }
```
[the remark this answers is not in the document]
Comment (d4vidi) [reply detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/uimodule/fabric/FabricUIManagerIdlingResources.kt:39]: Can fix?

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/timers/JavaTimersReflected.kt [hunk 13]
```diff
@@ -0,0 +1,26 @@
+package com.wix.detox.reactnative.idlingresources.timers
+
+import com.facebook.react.bridge.ReactContext
+import com.facebook.react.modules.core.JavaTimerManager
+import org.joor.Reflect
+import kotlin.reflect.full.declaredFunctions
+import kotlin.reflect.jvm.isAccessible
+
+object JavaTimersReflected {
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/timers/JavaTimersReflected.kt:9]: Cool

File: detox/scripts/updateGradle.js [hunk 14]
```diff
@@ -26,6 +26,35 @@ function getGradleVersionByRNVersion() {
...
+  }
+
+  if (!isNewArch) {
+    try {
+      const gradleProperties = path.join(process.cwd(), 'android', 'gradle.properties');
+      console.log(`Patching gradle.properties. File: ${gradleProperties}`);
+      const data = fs.readFileSync(gradleProperties, 'utf8');
+      const updatedData = data.replace('newArchEnabled=true', 'newArchEnabled=false');
+      fs.writeFileSync(gradleProperties, updatedData, 'utf8');
+      console.log('gradle.properties patched successfully.');
```
[the remark this answers is not in the document]
Comment (d4vidi) [reply detox/scripts/updateGradle.js:52]: WDYT? 😄

File: detox/android/detox/build.gradle (hunk 10 above)
Comment (gosha212) [reply detox/android/detox/build.gradle:155]: Google Chrome 2025-01-16 14 42 50

**Comment by mikehardy** [comment]
Hey @gosha212 - just wanted to say thanks for the new arch fixes in detox - I know it's a pain getting the ReactContext because I also had to fight my way through it in react-native-firebase and notifee, with all the symbols being different across versions but you need it to work on all the versions too, what a pain.

I tried the 20.33.0-prerelease.0 release and it works for me (at least with one trivial build.gradle patch that's already tracked here in an issue). 20.32.0 definitely didn't get the react context in new arch mode, but I've got all the correct adb logcat messages now indicating it has it with these PRs in the prerelease 🔥

**Comment by gosha212** [comment]
@mikehardy thanks for the warm words. The iOS is 99% ready

**Comment by mikehardy** [comment]
react-native-firebase e2e app (my use case here) does almost no app interaction so we are not a great indicator of full-feature functionality, but it's working for us on react-native iOS platform as well as react-native-macos already, for basic app launch / terminate / ready-signaling behavior, so I'm well-pleased. Merged the PR with it already and our e2e app is able to test (and sadly, uncover some) new architecture issues

**Comment by d4vidi** [comment]
@mikehardy thanks again for the kind words! 🙏🏻
On iOS, we've been finalizing some edge cases related to Detox's handling of app crash scenarios.

>...able to test (and sadly, uncover some) new architecture issues

Music to our ears... 🎶  whenever tests uncovers issue, we feel better about our work 😊

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/timers/JavaTimersReflected.kt [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+            hasActiveTimersInRangeInstanceClass.declaredFunctions.first { it.name.contains("hasActiveTimersInRange") }
+        method.isAccessible = true
+        val hasActiveTimers: Boolean = method.call(timersManager, BUSY_WINDOW_THRESHOLD) as? Boolean ?: false
+        return hasActiveTimers
+    }
+
+    private fun getTimersManager(reactContext: ReactContext): JavaTimerManager {
+        val reactHost = Reflect.on(reactContext).field("mReactHost").get<Any>()
+        val reactInstance = Reflect.on(reactHost).field("mReactInstance").get<Any>()
+        return Reflect.on(reactInstance).field("mJavaTimerManager").get() as JavaTimerManager
```
Comment (mikehardy) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/timers/JavaTimersReflected.kt:24]: Hey @gosha212 👋 

Saw two things here:

1- they converted this file to kotlin (first released in react-native 0.80) and it changed from `mJavaTimerManager` to the non-Hungarian notation `javaTimerManager`:

https://github.com/facebook/react-native/blame/0e6b94f4ab62a5cfa9cb770615e255fca31221f9/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/runtime/ReactInstance.kt#L101

Maybe 

```kotlin

        try {
          return Reflect.on(reactInstance).field("mJavaTimerManager").get() as JavaTimerManager
        } catch (nsfe: NoSuchMethodException) {
          return Reflect.on(reactInstance).field("javaTimerManager").get() as JavaTimerManager
        }
```

**Comment by gosha212** [comment]
@mikehardy hi, there is already a new pr to support rn 80
